lib/astutils.cpp

```diff
@@ -3647,3 +3647,82 @@ bool isUnevaluated(const Token *tok)
 {
     return Token::Match(tok, "alignof|_Alignof|_alignof|__alignof|__alignof__|decltype|offsetof|sizeof|typeid|typeof|__typeof__ (");
 }
+
+static std::set<std::string> getSwitchExprs(const Token *startbrace, bool &hasDefault)
+{
+    std::set<std::string> exprs;
+    const Token *endbrace = startbrace->link();
+    if (!endbrace)
+        return {};
+
+    hasDefault = false;
+    for (const Token *tok = startbrace->next(); tok && tok != endbrace; tok = tok->next()) {
+        if (Token::simpleMatch(tok, "{")) {
+            tok = tok->link();
+            continue;
+        }
+        if (Token::simpleMatch(tok, "case")) {
+            const Token *exprtok = tok->astOperand1();
+            if (!exprtok)
+                continue;
+            while (Token::Match(tok, "%name% ::"))
+                tok = tok->tokAt(2);
+            exprs.insert(exprtok->str());
+            continue;
+        }
+        if (Token::simpleMatch(tok, "default")) {
+            hasDefault = true;
+            continue;
+        }
+    }
+
+    return exprs;
+}
+
+bool isExhaustiveSwitch(const Token *startbrace)
+{
+    if (!startbrace)
+        return false;
+    const Token *rpar = startbrace->previous();
+    if (!rpar)
+        return false;
+    const Token *lpar = rpar->link();
+    if (!lpar)
+        return false;
+    const Token *vartok = lpar->next();
+    // bail out on more complex expression than just a variable
+    if (!vartok || vartok->next() != rpar || !vartok->isVariable())
+        return false;
+
+    bool hasDefault{};
+    std::set<std::string> caseExprs = getSwitchExprs(startbrace, hasDefault);
+
+    if (hasDefault)
+        return true;
+
+    if (vartok->variable()->getTypeName() == "bool") {
+        const auto &trueCase = caseExprs.find("true");
```

Comment:
as far as I know iterators are cheap to copy by value so a reference is not needed here.